---
title: "SpatialinferCNV"
author: "Yelin Zhao"
date: "2024-03-14"
output: html_document
---
https://aerickso.github.io/SpatialInferCNV/
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Install packages
```{r cars}
install.packages("devtools")
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("infercnv")
#install.packages("tidyverse")
#install.packages("Seurat")
install.packages("phylogram")
install.packages("ape")
install.packages("hdf5r")
install.packages("rjags")
library(devtools)
install_github("aerickso/SpatialInferCNV")

```

#Load packages
```{r cars}
library(Seurat)
library(SpatialInferCNV)
library(tidyverse)
library(tibble)
library(dplyr)
library(stringr)
```

# Downloading and importing publicly available Visium Datasets (10x Genomics)
The 10X Genomics SpaceRanger Pipeline outputs, among other files, filtered_feature_bc_matrix.h5 files that store count data. Count data, such as from the Visium assay, is a key input to Spatial InferCNV. 10X Genomics offers a number of publicly available 10x Genomics Visium datasets https://www.10xgenomics.com/datasets?query=&page=1&configure%5BhitsPerPage%5D=50&configure%5BmaxValuesPerFacet%5D=1000.

Start by downloading a publicly available 10x Genomics Visium dataset Breast Cancerhttps://www.10xgenomics.com/datasets/human-breast-cancer-block-a-section-1-1-standard-1-1-0.

Then, read in the filtered_feature_bc_matrix.h5 file and convert it to the R data.frame format. Additionally, we will append a “Section Label”. This is because each 10x Genomics Visium assay uses the same set of barcodes, and thus we need to append an section identifier.

```{r}
# Downloading Breast Cancer Data H5 file
download.file("https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Breast_Cancer_Block_A_Section_1/V1_Breast_Cancer_Block_A_Section_1_filtered_feature_bc_matrix.h5", "/SpatialInferCNV/10x breastcancer/V1_Breast_Cancer_Block_A_Section_1_filtered_feature_bc_matrix.h5", mode = "wb")

# Reading the 10X Visium Breast Cancer count data, and appending the string "Breast10X" before each barcode with ImportCountData()
Breast_ENSBMLID_Counts <- ImportCountData("Breast10X", "/Users/yelin.zhao/Library/CloudStorage/OneDrive-Personal/Projects/Spatial/SpatialInferCNV/10x breastcancer/V1_Breast_Cancer_Block_A_Section_1_filtered_feature_bc_matrix.h5")

# Showing the first part of the Breast_ENSBMLID_Counts dataframe 
head(Breast_ENSBMLID_Counts)[,1:3]
```

#Importing Histological Annotation Data
Next, the user will import the histological annotation .csv file containing histological annotations and convert it to the R data.frame format. This was created in the previous step, but we also provide this .csv file in the directory if you wish to skip to this step.

The motivations for annotations is described above and in our manuscript. These annotations are typically used in analyses to subselect or filter for certain histological populations.
```{r}
Userguide_10xBreast_Histology <- ImportHistologicalAnnotations("Breast10X", "/Users/yelin.zhao/Library/CloudStorage/OneDrive-Personal/Projects/Spatial/SpatialInferCNV/10x breastcancer/10xBreast_UserguideHistologyAnnotations.csv")
head(Userguide_10xBreast_Histology)
```
#Merging histological annotation data with count data
This next step demonstrates functionality of MergingCountAndAnnotationData(). This function requires 3 parameters: a section annotation string (this needs to be identical from the annotation and count files), the annotation data.frame, and the count data.frame)

This step will also apply a QC filter, to remove any barcodes that fall below the recommended QC threshold of a minimum of 500 counts.
```{r}
# Merging annotations, and counts, and applying QC with MergingCountAndAnnotationData()
Userguide_10xBreast_Joined_Counts <- MergingCountAndAnnotationData("Breast10X",Userguide_10xBreast_Histology, Breast_ENSBMLID_Counts)

# Setting the "Genes" column as the rownames
Userguide_10xBreast_Joined_Counts <- Userguide_10xBreast_Joined_Counts %>% column_to_rownames(var = "Genes")

# Removing the pre-merged count file as it is no longer needed
rm(Breast_ENSBMLID_Counts)

# Showing the first part of the Userguide_10xBreast_Histology dataframe 
head(Userguide_10xBreast_Joined_Counts)[,-1:-3]

```

#Selecting finalized annotations for export for use in infercnv::run
During pre-processing, you may filter out certain barcodes (either due to selection of certain histological populations, or due to not passing QC). The function FinalAnnotations() compares the processed count file (which has all of the count data to be analyzed) with the original parent histological annotation file, removes any extra annotations not included in the count file, and returns a finalized annotation dataframe for export.

In this example, we will simulate this, by dropping the first barcode from the count file processed above.
```{r}
#Diplaying the dimensions of the original annotation file: 12 rows, 2 columns
dim(Userguide_10xBreast_Histology)

#Dropping the first barcode count data
Userguide_10xBreast_Joined_Counts <- Userguide_10xBreast_Joined_Counts %>% select(-Breast10X_AATAACGTCGCGCCCA.1)

#Running FinalAnnotations() to return only the barcodes in the count file
FinalAnnotationsForExport <- FinalAnnotations(Userguide_10xBreast_Histology, Userguide_10xBreast_Joined_Counts)

#Diplaying the dimensions of the final annotation: 11 rows, 2 columns
dim(FinalAnnotationsForExport)
```

#Manual selection of output groups from a infercnv run plot for visualization and further analysis
In order to run the next function, the user needs to run infercnv. InferCNV requires 3 inputs: a count file, an annotation file, and a gene loci file. The code will output the first annotation and count file created above. Documentation to create a gene loci file can be found here at the infercnv wiki documentation https://github.com/broadinstitute/inferCNV/wiki/instructions-create-genome-position-file. To save you time, we provide a file download in the chunk. Author Note: given that the repo is private, this is currently commented out: manually download the siCNV_GeneOrderFile.tsv file from the repo.
```{r}
# Download the siCNV_GeneOrderFile to the local directory

# Author Note: given that the repo is private, this is currently commented out: manually download the file from the repo
download.file("https://raw.githubusercontent.com/aerickso/SpatialInferCNV/main/FigureScripts/siCNV_GeneOrderFile.tsv?token=GHSAT0AAAAAABRUPXQSKNJYB6FRBX2GLUGMYR77RCQ", "./10x breastcancer/siCNV_GeneOrderFile.tsv", mode = "wb")

# Write the Userguide_10xBreast_Joined_Counts.tsv to the local directory
write.table(Userguide_10xBreast_Joined_Counts, "./10x breastcancer/Userguide_10xBreast_Joined_Counts.tsv", sep = "\t")

# Write the Userguide_10xBreastFinalAnnotationsForExport.tsv to the local directory
write.table(FinalAnnotationsForExport, "./10x breastcancer/Userguide_10xBreastFinalAnnotationsForExport.tsv", 
            sep = "\t",
            quote = FALSE, 
            col.names = FALSE, 
            row.names = FALSE)

# Create the inferCNV object from the 3 files, without the reference group and mitochondria 
Userguide_10xBreastCancer_infCNV <- infercnv::CreateInfercnvObject(raw_counts_matrix="./10x breastcancer/Userguide_10xBreast_Joined_Counts.tsv", 
                                               gene_order_file="./10x breastcancer/siCNV_GeneOrderFile.tsv",
                                               annotations_file="./10x breastcancer/Userguide_10xBreastFinalAnnotationsForExport.tsv",
                                               delim="\t",
                                               ref_group_names=NULL, 
                                               chr_exclude = c("chrM")) 

# Run inferCNV
Userguide_10xBreastCancer_infCNV = infercnv::run(Userguide_10xBreastCancer_infCNV, 
                                              cutoff=0.001, #(see infercnv::run documentation)
                                              out_dir="./10x breastcancer/InferCNVrun_outputs", 
                                              cluster_by_groups=FALSE, #unsupervised analysis
                                              HMM = FALSE, 
                                              denoise=TRUE) #denoising applies noise reduction for the plot 
```
This is where SpatialInferCNV’s SelectingSubTreeData() function can help: it allows you to identify barcodes associated with certain clusters by manual selection from the dendrogram at left. For this The dendrogram, visualized at left in the above denoised output plot, is also generated in infercnv::run, and can be found at “./InferCNVrun_outputs/infercnv.21_denoised.observations_dendrogram.txt”.

Please note: this next step requires manual interpretation of the images (often done in an image viewer or editor outside of R).

For the purposes of the userguide, lets assume that we have identified signal in certain groups of Visium spots (here termed “clones”). Lets proceed with about extracting the identities of the barcodes for clone 1, 2 and 3 highlighted in boxes the following image (edited in Microsoft Paint).
```{r}
library(ape)
library(phylogram)

#Use read.dendrogram() to import the dendogram file
BreastCancer10x_for_clustering <- read.dendrogram(file = "./UserGuideFiles/infercnv.21_denoised.observations_dendrogram.txt")

#Convert to a phylo object using as.phylo()
BreastCancer10x_for_clustering_phylo <- as.phylo(BreastCancer10x_for_clustering)

#Use subtrees() to enable further interaction with the dendrogram
my.subtrees = subtrees(BreastCancer10x_for_clustering_phylo)  # subtrees() to subset

#Output an image to visualize all of the dengdrogram nodes 
png("BreastCancer10x_forclustering_phylo.png",width=10000,height=2500, res = 300)
plot(BreastCancer10x_for_clustering_phylo,show.tip.label = FALSE)
nodelabels(text=1:BreastCancer10x_for_clustering_phylo$Nnode,node=1:BreastCancer10x_for_clustering_phylo$Nnode+Ntip(BreastCancer10x_for_clustering_phylo))
dev.off()
```

If we manually inspect the image above, we can identify the dendrogram nodes associated with the clone groups of interest.
We then select these nodes data using SelectingSubTreeData(), with the my.subtrees object created above, and the numerical node number in the output image. The output of SelectingSubTreeData() is a data.frame, with barcodes in the first column, and the node number (identified above), in the second column.
```{r}
#Selecting Node 9
Node9 <- SelectingSubTreeData(my.subtrees, 9)

#Selecting Node 5
Node5 <- SelectingSubTreeData(my.subtrees, 5)

#Selecting Node 3
Node3 <- SelectingSubTreeData(my.subtrees, 3)

#Merging All Nodes Together
MergedNodes <- rbind(Node5, Node9)
MergedNodes <- rbind(MergedNodes, Node3)

#Note, since the "additional spot" for Clone 3 did not have a node, identify it by joining the original annotations to the MergedNodes dataframe
MergedNodes <- full_join(MergedNodes, FinalAnnotationsForExport)

#The additionally joined barcode does not have a node identity, so manually change it to "Clone 3" (in line with the image above)
MergedNodes$Node <- ifelse(is.na(MergedNodes$Node), "Clone 3", MergedNodes$Node)

#Then drop the Histology column
MergedNodes <- MergedNodes %>% select(-Histology)

#This then renames "Node" to "Clone"
MergedNodes$Node <- ifelse(MergedNodes$Node == "Node_9", "Clone1",
                     ifelse(MergedNodes$Node == "Node_5" , "Clone2",
                     ifelse(MergedNodes$Node == "Node_3" , "Clone2", MergedNodes$Node)))

#Show the first part of the finalized dataframe for this step
head(MergedNodes)

```

The MergedNodes data.frame object can be used, for example, to output a new annotation file, to be imported into inferCNV, and then run inferCNV with cluster_by_groups=TRUE, to output a plot, clustered by clones.

The identified clones now can also be visualized in the LoupeBrowser for analysis or generation of figure images. To do so, they need to be exported as a .csv file, and the “section name” that we appended to the barcodes needs to be removed.

```{r}
#Copy the MergedNodes dataframe to a new dataframe called "ForLoupeBrowser"
ForLoupeBrowser <- MergedNodes

#Use the function substr() to remove the first 9 characters in the barcode column (removes "Breast10X_")
ForLoupeBrowser$Barcode <- trimws(substr(ForLoupeBrowser$Barcode, 11, 100))

#Replace the "." in the current barcode dataframe to "-" (required for LoupeBrowser)
ForLoupeBrowser$Barcode <- gsub("\\.", "\\-", ForLoupeBrowser$Barcode)

#Write the dataframe to csv
write.csv(ForLoupeBrowser, "Userguide_BreastCancer_Clones_ForLoupeBrowser.csv", row.names = FALSE)
```






# My old code for inferCNV
```{r}
#InferCNV
# module load R/4.0.4
# library("devtools")
# devtools::install_github("broadinstitute/infercnv")
rm(list=ls())
myPaths <- .libPaths() 
# myPaths <- c(myPaths, "/home/yelzh67/R/x86_64-pc-linux-gnu-library/4.0") # add new path
myPaths <- c(myPaths[2], myPaths[1])  # switch them
.libPaths(myPaths)  

# remotes::install_version("Seurat", version = "4.0.4")
library(Seurat)
library(clusterProfiler)  
library(DOSE)
library(topGO)
library(pathview)
library(dplyr)
library(ggplot2)
library(patchwork)
library(RColorBrewer)
library(infercnv)
# install.packages('AnnoProbe')
library(AnnoProbe)


location = "omika"
cancer_name = c("Lung","Liver","Colon","Ovary","Breast")

if (location == "mac"){  wd="/Users/cynthia_ye/Documents/OneDrive - Linköpings universitet"}
if (location == "omika"){ wd="/home/yelzh67/Projects"}

inputdir = paste0(wd,"/cancer/datafile/outputs_5cancers/new_outputs/Seurat_subset_celltyping/2022Feb12")
outdir= paste0(wd,"/cancer/datafile/outputs_5cancers/new_outputs/InferCNV/20220803-onlyepi")
if (dir.exists(outdir)==F){dir.create(outdir)
  print("outdir created")} else {print("outdir existed")}

################################################################
################################################################
i=1 done
i=2 
i=3 done
i=4 done
4:length(cancer_name)

for (i in c(2,5)){
  out.dir= paste0(outdir,"/",cancer_name[i])
  if (dir.exists(out.dir)==F){dir.create(out.dir)
    print("out.dir created")} else {print("out.dir existed")}
  
  #1. Prepare files####
  load(paste0(inputdir,"/all_integrated_Celltype_final_",cancer_name[i],".RData"))
  all.integrated@meta.data %>% head
  table(all.integrated@meta.data$Celltype_final)
  all.integrated@meta.data$celltypeforInferCNV = paste0(all.integrated@meta.data$group,"_",all.integrated@meta.data$Celltype_final)
  table(all.integrated@meta.data$celltypeforInferCNV)
  
  all.integrated =  subset(x = all.integrated, 
                           subset = celltypeforInferCNV %in% c("T_Epithelial","N_Epithelial")) #,"N_Epithelial","N_Fibroblast","T_Fibroblast","T_Endothelial"
  ## random subset 5000 cells if nrow(all.integrated@meta.data) >5000
  if (nrow(all.integrated@meta.data) >5000) {
    print("N cells >5000, randomly subset 5000 cells")
    all.integrated = subset(all.integrated, downsample = 5000)
  }
  ## get the matrix file
  matrix_file = as.data.frame(GetAssayData(all.integrated , slot='counts',assay='RNA'))
  matrix_file[1:5,1:5]
  
  groupinfo = data.frame(v1=colnames(all.integrated),
                         v2=all.integrated@meta.data$celltypeforInferCNV)
  table(groupinfo$v2,useNA =  "always")
  groupinfo[is.na(groupinfo$v2),]
  identical(colnames(matrix_file),groupinfo$v1)
  
  geneInfor=annoGene(rownames(all.integrated),"SYMBOL",'human')
  colnames(geneInfor)
  geneInfor=geneInfor[with(geneInfor, order(chr, start)),c(1,4:6)]
  geneInfor=geneInfor[!duplicated(geneInfor[,1]),]
  length(unique(geneInfor[,1]))
  head(geneInfor)
  dim(geneInfor)
  which(is.na(geneInfor$chr))
  geneInfor = geneInfor[is.na(geneInfor$chr) == FALSE,] #remove gene which the chromosomes is not properly assigned
  geneInfor = geneInfor[(geneInfor$chr == "chrM") == FALSE,] #remove gene which the chromosomesM
  geneInfor = geneInfor[(geneInfor$chr == "chrX") == FALSE,] #remove gene which the chromosomesX
  geneInfor = geneInfor[(geneInfor$chr == "chrY") == FALSE,] #remove gene which the chromosomesY
  
  # Make sure that chromosomes are ordered correctly 
  geneInfor$chr = factor(geneInfor$chr, 
                       levels = c("chr1", "chr2","chr3","chr4", "chr5", "chr6","chr7", "chr8", "chr9","chr10", "chr11", "chr12","chr13", "chr14", "chr15","chr16", "chr17", "chr18","chr19", "chr20", "chr21","chr22"))
  geneInfor = geneInfor[order(geneInfor$chr),]
  
  expFile=paste0(out.dir,'/expFile.txt')
  write.table(matrix_file,file = expFile,sep = '\t',quote = F,col.names = T,row.names = T)
  groupFiles=paste0(out.dir,'/groupFiles.txt')
  write.table(groupinfo,file = groupFiles,sep = '\t',quote = F,col.names = F,row.names = F)
  geneFile=paste0(out.dir,'/geneFile.txt')
  write.table(geneInfor,file = geneFile,sep = '\t',quote = F,col.names = F,row.names = F)
  
  ################################################################
  #2. Run InferCNV####
  options(stringsAsFactors = F)
  expFile=paste0(out.dir,'/expFile.txt')
  groupFiles=paste0(out.dir,'/groupFiles.txt')
  geneFile=paste0(out.dir,'/geneFile.txt')
  
  infercnv_obj = CreateInfercnvObject(raw_counts_matrix=expFile,
                                      annotations_file=groupFiles,
                                      delim="\t",
                                      gene_order_file= geneFile,
                                      ref_group_names=c("N_Epithelial")) #"N_Epithelial(non-malignant)",'T_Endothelial','T_Fibroblast',,"T_Fibroblast","T_Endothelial"
  unique(infercnv_obj@gene_order$chr)
  table(infercnv_obj@gene_order$chr)

  # Make sure that chrmosomes are ordered correctly 
  slot(infercnv_obj, "gene_order")[,"chr"] <- factor(slot(infercnv_obj, "gene_order")[,"chr"], 
                                                     levels = c("chr1", "chr2","chr3","chr4", "chr5", "chr6","chr7", "chr8", "chr9","chr10", "chr11", "chr12","chr13", "chr14", "chr15","chr16", "chr17", "chr18","chr19", "chr20", "chr21","chr22","chrX"))
   
  # perform infercnv operations to reveal cnv signal
  infercnv_obj = infercnv::run(infercnv_obj,
                               cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics
                               out_dir=out.dir,  # dir is auto-created for storing outputs
                               cluster_by_groups=T,   # perform separate clustering for the tumor cells according to the patient type, as defined in the cell annotations file.
                               denoise=T,
                               cluster_references = T,
                               analysis_mode = "subclusters",
                               HMM=T,
                               k_obs_groups=5 ,
                               num_threads=20,
                               output_format="pdf")
  
  # infercnv_obj = readRDS('preliminary.infercnv_obj')
  # infercnv::plot_cnv(infercnv_obj, #上两步得到的infercnv对象
  #                    plot_chr_scale = T, #画染色体全长，默认只画出（分析用到的）基因
  #                    out_dir = out.dir,
  #                    output_filename = "better_plot",output_format = "pdf" ) #保存为pdf文件
  #                    # custom_color_pal =  color.palette(c("#8DD3C7","white","#BC80BD"), c(2, 2)) #改颜色
  # 
}




################################################################
```









